src/sas/qtgui/MainWindow/DataExplorer.py

```diff
@@ -534,6 +535,47 @@ def readProject(self, filename):
         self.cbFitting.setCurrentIndex(
                 self.cbFitting.findText(visible_perspective))
 
+    def readConstraintsFromProject(self, all_data):
+        """
+        Extracts constraints from *all_data* dict and returns a dict where keys
+        are the tab name and values are a list of constraints on that tab. The
+        dict can then be passed to the updateFromConstraints method from the
+        fitting perspective to apply the constraints with error checking
+        mechanism
+        """
+        constraint_dict = {}
+        for key, value in all_data.items():
+            # make sure we dealing with params
+            if 'fit_params' not in value:
+                continue
+            params = value['fit_params']
+            for page in params:
+                if not isinstance(page, dict):
+                    continue
+
+                tab_name = None
+                constraint_list = []
+                for param_name, param_value in page.items():
+                    # make sure we are dealing with lists
+                    if not isinstance(param_value, list):
+                        continue
+                    # get the tab name
+                    elif param_name == "tab_name":
+                        tab_name = param_value[0]
+                    # get parameters
+                    elif len(param_value) >= 5:
+                        ioffset = 1 if len(param_value) > 5 else 0
+                        cons = param_value[4+ioffset]
+                        # get the constraint
+                        if cons is not None and len(cons) == 5:
+                            constraint_list.append(cons)
+                    else:
+                        continue
+
+                if tab_name and constraint_list:
+                    constraint_dict.update({tab_name: constraint_list})
+        return constraint_dict
```

Comment:
The `readConstraintsFromProject` function belongs with the data reader (currently in `GuiUtil`) since it will also be useful when running fits headless. It does not need to be a method in this class since it makes no reference to `self`. 

It may make sense to have it as a method on the project data structure, but this is a style preference: `data.get_constraints()` vs. `get_constraints(data)`. Don't make it an attribute like `data.constraints`: it does more work than people expect when accessing an attribute, and it won't be used frequently enough to justify writing cacheing code to make it efficient. Don't call it `read` since it isn't actually reading the file.

You may want to move some of the checking directly into the project reader so that the data structures are clean and this code can be made simpler.  In general cleaning up data on input saves trouble in the rest of the code (for example, I'm guessing that the length greater than 5 test on line 567 shows up elsewhere), but if you hide the internals through method calls like `get_constraints()` there won't be much difference.

There may be some advantage to collecting and reporting errors and warnings on load before the user does too much work with the loaded data, but this particular code doesn't do any error reporting (whether it should be reporting errors is a separate question).

Reply:
... make that `fit_project.get_constraints()` or `get_constraints(fit_project)` since a fit project may contain multiple data sets.